tests/core/data/test_network.py

```diff
@@ -0,0 +1,334 @@
+import numpy as np
+from numpy.testing import assert_almost_equal
+import pytest
+
+from qutip.core import data
+from qutip.core.data import dense
+from qutip_tensornetwork.core.data import Network
+from qutip_tensornetwork.testing import assert_network_close
+from .conftest import random_node, random_complex_network
+import tensornetwork as tn
+
+def test_init_only_in_edge():
+    node = random_node((2,))
+    network = Network([], node[0:], copy=False)
+    assert len(network.nodes) == 1
+    assert network.nodes.pop() is node
+    assert len(network.in_edges) == 1
+    assert network.in_edges[0] is node[0]
+    assert len(network.out_edges) == 0
+
+def test_init_only_out_edge():
+    node = random_node((2,))
+    network = Network(node[0:], [], copy=False)
+    assert len(network.nodes) == 1
+    assert network.nodes.pop() is node
+    assert len(network.out_edges) == 1
+    assert network.out_edges[0] is node[0]
+    assert len(network.in_edges) == 0
+
+def test_init_both_in_and_out_edge():
+    node = random_node((2, 2))
+    network = Network(node[0:1], node[1:], copy=False)
+    assert len(network.nodes) == 1
+    assert network.nodes.pop() is node
+    assert len(network.out_edges) == 1
+    assert network.out_edges[0] is node[0]
+    assert len(network.in_edges) == 1
+    assert network.in_edges[0] is node[1]
+
+def test_init_raise_if_edges_not_unique():
+    """in_edges and out_edges must be unique."""
+    node = random_node((2, 2))
+    with pytest.raises(ValueError):
+       network = Network([node[0]], [node[0]])
+
+def test_non_dangling_in_or_out_raises():
+    """Input edges for in_edges and out_edges must be dangling."""
+    node = random_node((2,))
+    node2 = random_node((2,))
+    node[0] ^ node2[0]
+
+    with pytest.raises(ValueError):
```

Comment:
Could we also test that these raise suitable error messages?

Reply:
I am actually againts testingt the error message itself. I do see why testing the error type raised is important. Including the error type in the tests ensures that in the future this same function will continue raising the same error. This is important because a user may catch this specific error as part of its program's logic and run into problems if the error is not `ValueError`.  However, I think the user is not supposed to use the string from the error as part of its program logic. It is meant to be a clarification of what happened and how to use the class. In this sense, I see the error message fulfilling the same role as the documentation, which is not included in the tests either.

Reply:
Error messages are not documentation though -- they're generated by the program and so can have bugs that are not immediately obvious. E.g. I've often had cases where `raise ValueError(f"{x} must be less than {y}")` should have been `raise ValueError(f"{y} must be less than {x}")` or the error message is just nonsense or not easily readable and tests are the only way to catch those.

Yes, code should not generally depend on the error string itself, but it's fine for tests to depend on the error string if the string is generated by the code base being tested. While code shouldn't depend on the error string, humans do when they're debugging.

There are also other cases where serious bugs can result from error messages -- e.g. `raise ValueError(f"password {p} is invalid")` which is a common security hole. Luckily we don't encounter these kinds of cases too often in QuTiP.